### PR TITLE
Add AES-CTR, make Algo configurable

### PR DESCRIPTION
**Pull request by kojo1**

### CODE REVIEW

**Comment by ejohnstown** [comment]
I found it the problem. wolfCrypt's AES-CTR mode doesn't work when in and out are the same. It encrypts the counter into out, then it XORs in into out. If in and out are the same, you overwrite in with the encrypted counter, then XOR out and in together, which are the same, you get zeroes. I'm going to submit a PR over on wolfSSL to change this to use an intermediate temp.

**Comment by kojo1** [comment]
Hi John,

Thank you for looking into and taking care.

Kojo



2019年4月4日(木) 6:56 John Safranek <notifications@github.com>:

**Comment by kojo1** [comment]
Thank you for taking care.
Meanwhile, made cannedXXXAlgoNames configurable.

**Comment by ejohnstown** [comment]
Jenkins please retest

**Comment by ejohnstown** [comment]
retest this please

**Comment by ejohnstown** [comment]
1. Add `#define WOLFSSL_AESCTR` to the file ide/winvs/user_settings.h.
2. Add `#ifdef WOLFSSL_AESCTR` guards around the calls to wc_AesCtrEncrypt().
3. Also add the guard around the AES-CTR cases in SendNewKeys() and DoNewKeys().

**Comment by kojo1** [comment]
Would they be ok with WOLFSSH_NO_xxx? Would the names be aligned with WOLFSSL_xxx?
VS build is not working.

File: ide/winvs/user_settings.h [hunk 1]
```diff
@@ -11,7 +11,7 @@
 #define HAVE_ECC
 #define HAVE_AESGCM
 #define HAVE_HASHDRBG
-#define WOLFSSL_AESCTR
+#define WOLFSSL_AES_COUNTER
```
Comment (ejohnstown) [inline ide/winvs/user_settings.h]: Oh yeah, the flag is `WOLFSSL_AES_COUNTER`. I'm sorry for the wrong request.

**Comment by ejohnstown** [comment]
> Would they be ok with WOLFSSH_NO_xxx? Would the names be aligned with WOLFSSL_xxx?
> VS build is not working.

They should be referring to the wolfCrypt configure options, which start with WOLFSSL. For example, for the key generation functions we check for WOLFSSL_KEY_GEN. The wolfSSL/wolfCrypt configuration does guide the wolfSSH configuration.

File: ide/winvs/user_settings.h (hunk 1 above)
Comment (kojo1) [reply ide/winvs/user_settings.h]: But still not seem to work. Don't see why.

**Pull request merged**